BUILDING.md

```diff
@@ -0,0 +1,578 @@
+
+This make file can:
+
+- run on GNU and BSD systems (many commands have flags set by variables)
+- be run with different versions of Make (not just GNU make)
+- cross-compile cockroach to darwin, linux-gnu, linux-musl, and windows
+
+
+### Approximate Sections of the Makefile
+
+Functionality is mostly split into relevant sections. These are the approximate line numbers for
+each section (they will surely change). Additonal details on each section are included below.
+
+- Initialization: 0 - 374
+- C deps: about 375 - 785
+- Cockroach: 801 - 965
+- Install/start cockroach: 966 - 976
+- Cockroach testing: 977 - 1075
+- Auxiliary commands: 1076 - 1159
+- .buildinfo: 1160 - 1183
+- Protobuf configuration: 1184 - 1307
+- UI: 1308 - 1414
+- SQL parser: 1415 - 1587
+- c-deps-fmt: 1588 - 1592
+- Cleaning: 1593 - 1638
+- Go binary dependency file generation: 1639 - 1732
+- Check CLI variables are all valid: 1733 - EOF
+
+
+#### Initialization Steps
+
+1. The `GO` and `xgo` (needed for `defs.mk`) variables are set.
+
+1. Some variables need to be computed using shell commands. The first step in the make file is to compute and cache a set of them in the `build/defs.mk` file so they don't get recomputed on every run of make. `build/defs.mk` is then included in `Makefile`.
+
+1. Many targets require the vendor modules to be set up. The second step that is performed is to set up the git submodules using the `bin/.submodules-initialized` target. The file is empty and only serves as a flag. The target is usually referenced as an order-only-prerequisite dependency (i.e., after the `|` on a target definition line).
+
+1. Sometimes you may want to set a variable the same for all runs of make. Add the variable(s) to the `customenv.mk` file and it will be used for every run of `make`. Format: `VAR_NAME = VALUE`.
+
+1. A check is performed to ensure `BENCHES` is only used with the `bench` goal and `TESTS` is not used with the `bench` goal.
+
+1. A check is performed to prevent invoking make with a specific test / bench name without specifying a constraining package.
+
+1. A check to ensure `TYPE` is not set with instructions to use `build/builder.sh mkrelease` instead.
+
+1. Building with dependency files (the `build-with-dep-files` variable) is set for all goals except `help` and `clean`.
+
+1. A large set of variables are defined, many with documentation, some with default values.
+
+    - documented: `PKG`, `TESTS`, `BENCHES`, `FILES`, `TESTCONFIG`, `SUBTESTS`, `LINTTIMEOUT`, `TESTTIMEOUT`, `RACETIMEOUT`, `ACCEPTANCETIMEOUT`, `BENCHTIMEOUT`, `TESTFLAGS`, `GOTESTFLAGS`, `STRESSFLAGS`, `CLUSTER`, `VERBOSE`, `DESTDIR`
+    - undocumented: `DUPLFLAGS`, `GOFLAGS`, `TAGS`, `ARCHIVE`, `STARTFLAGS`, `BUILDTARGET`, `SUFFIX`, `INSTALL`, `prefix`, `bindir`
+
+1. If the make command line contains the `-j` flag and NCPUS is set, then `-j${NCPUS}` is added to MAKEFLAGS.
+
+1. The `help` target is defined
+
+1. Many more variables are defined:
```

Comment:
I don't want to be a broken record, but I'm trying to understand who is this kind of thing helping? Why aren't these comments in the Makefile, for example? Did you find these variables in the Makefile perplexing, and that's why you added to this document?